src/enumeration.rs

```diff
@@ -424,3 +424,320 @@ impl std::fmt::Debug for InterfaceInfo {
+            .field("revision", &self.revision)
+            .field("subsystem_vendor_id", &self.subsystem_vendor_id)
+            .field("subsystem_device_id", &self.subsystem_device_id);
+
+        #[cfg(target_os = "linux")]
+        {
+            s.field("sysfs_path", &self.path);
+        }
+
+        #[cfg(target_os = "windows")]
+        {
+            s.field("instance_id", &self.instance_id);
+        }
+
+        s.finish()
+    }
+}
+
+/// USB host controller type
+#[derive(Copy, Clone, Eq, PartialOrd, Ord, PartialEq, Hash, Debug)]
+#[non_exhaustive]
+pub enum UsbController {
+    /// xHCI controller (USB 3.0+)
+    XHCI,
+
+    /// EHCI controller (USB 2.0)
+    EHCI,
+
+    /// OHCI controller (USB 1.1)
+    OHCI,
+
+    /// VHCI controller (virtual internal USB)
+    VHCI,
+}
+
+impl UsbController {
+    #[allow(dead_code)] // not used on all platforms
+    pub(crate) fn from_str(s: &str) -> Option<Self> {
+        match s
+            .find("HCI")
+            .filter(|i| *i > 0)
+            .and_then(|i| s.bytes().nth(i - 1))
+        {
+            Some(b'x') | Some(b'X') => Some(UsbController::XHCI),
+            Some(b'e') | Some(b'E') => Some(UsbController::EHCI),
+            Some(b'o') | Some(b'O') => Some(UsbController::OHCI),
+            Some(b'v') | Some(b'V') => Some(UsbController::VHCI),
+            _ => None,
+        }
+    }
+}
+
+/// Information about a system USB bus. Aims to be a more useful and portable version of the Linux root hub device.
+///
+/// Platform-specific fields:
+/// * Linux: `path`, `busnum`, `root_hub`
+/// * Windows: `instance_id`, `location_paths`, `devinst`, `driver`
+/// * macOS: `registry_id`, `location_id`, `name`, `driver`
+pub struct BusInfo {
+    #[cfg(target_os = "linux")]
+    pub(crate) path: SysfsPath,
+
+    /// The phony root hub device
+    #[cfg(target_os = "linux")]
+    pub(crate) root_hub: DeviceInfo,
+
+    #[cfg(target_os = "linux")]
+    pub(crate) busnum: u8,
+
+    #[cfg(target_os = "windows")]
+    pub(crate) instance_id: OsString,
+
+    #[cfg(target_os = "windows")]
+    pub(crate) location_paths: Vec<OsString>,
+
+    #[cfg(target_os = "windows")]
+    pub(crate) devinst: crate::platform::DevInst,
+
+    #[cfg(any(target_os = "windows", target_os = "macos"))]
+    pub(crate) driver: Option<String>,
+
+    #[cfg(target_os = "macos")]
+    pub(crate) registry_id: u64,
+
+    #[cfg(target_os = "macos")]
+    pub(crate) location_id: u32,
+
+    #[cfg(target_os = "macos")]
+    pub(crate) name: Option<String>,
+
+    /// System ID for the bus
+    pub(crate) bus_id: String,
+
+    /// Optional PCI information if the bus is connected to a PCI Host Controller
+    pub(crate) pci_info: Option<PciInfo>,
+
+    /// Detected USB controller type
+    pub(crate) controller: Option<UsbController>,
+
+    /// System provider class name for the bus
+    pub(crate) provider_class: Option<String>,
+    /// System class name for the bus
+    pub(crate) class_name: Option<String>,
```

Comment:
I'd be wary of getting into the business of generating human-readable text, because downstream application authors are never going to all agree on exactly what that text should look like, and that's before you even get into the can of worms that is localization.

We're already in the situation of having to consider parsing some of these details out of strings "helpfully" generated by the OS; it'd be good not to add another layer of that on top.

I think it's better to just stick to just reporting the known facts: this is bus `N` on an `XHCI` controller handled by driver `Y`, etc, and then everyone's free to generate text in any language to suit their preferences from the facts available.

Reply:
I had this conflict myself too, all good points. I went for the middle ground of:

* Separating `class_name`, `provider_class_name` to macOS with accessor methods.
* Adding the field for `root_hub_description` on Windows.
* Made `driver` universal since it can be read from the pci driver symblic link filename. This is actually better than relying on the sysfs manufacturer field I think.
* Created a `system_name()` (maybe better as just `name()`?) accessor method for all platforms, which returns `name` on macOS, `root_hub_description` on Windows and the root hub product string on Linux.